doc/source/Versioning.rst

```diff
@@ -0,0 +1,134 @@
+  2. It's a way to announce the latest changes to the library since the last milestone.
+
+Currently, the repository tracks 2 separate version numbers: the primary version number (used by the documentation and the shared/static library) **and** the pygrackle version number.
+
+.. _primary-version-number:
+
+Primary Version Number
+----------------------
+
+The version number used in a release has the form ``<MAJOR>.<MINOR>(.<MICRO>)``, and the version number tracked in the repository at any given time follows the more general form ``<MAJOR>.<MINOR>(.<MICRO>)(.dev<DEV_NUM>)``.
+In this template ``<MAJOR>``, ``<MINOR>``, and ``<MICRO>`` correspond to major, minor, and micro version numbers (the micro version number may be omitted if it’s zero). The final section can specify a development version.
+
+Version numbers associated with a release follow the rules of `semantic versioning <https://semver.org/>`__. [#f1]_
+
+
+.. COMMENT BLOCK
+
+   If we were following the rules of semantic versioning as strictly
+   as possible, we would NEVER remove deprecated functionality except
+   when incrementing the major version number.  For example, in the
+   release of Grackle 3.0, we marked some functions as deprecated.
+   Technically, if we were adhearing to the rules of semantic
+   versioning as strictly as possible, we would have waited until
+   Grackle 4.0 to actually remove those functions.  The main reason to
+   do this is if we cared about maintaining ABI compatability between
+   minor versions (it would make the shared library numbering make
+   more sense)
+
+   With all of that said, it's still technically correct for us to say
+   that we follow the semantic versioning specification. An explicit
+   exception appears to be carved out for removing deprecated
+   functionality.
+
+A change in the major version number typicially denotes significant changes in functionality.
+It must also be incremented in a new release if a backwards incompatible change has been made to the public API since the last release (alternatively, parts of the public API can be deprecated in a major release and be removed later on in a minor release).
+
+A change in the minor version number can indicate changes in functionality.
+It must be incremented in a new release if a backwards-compatable change has been made to the public API since the last release.
+
+The presence of ``.dev<DEV_NUM>`` in a version number denotes a development version.
+Development version numbers are not compatible with semantic versioning.
+
+For context, the first 10 version numbers since version ``3.0``, in the order of release are: ``3.0``, ``3.1.dev1``, ``3.1``, ``3.2.dev1``, ``3.1.1``, ``3.2.dev2``, ``3.2.0``, ``3.3.dev0``, ``3.2.1``, ``3.3.dev1``
+
+.. COMMENT BLOCK
+
+   There are 2 reasons that development versions are not compatible
+   with semantic versioning:
+
+   1. From a formatting perspective, the delimiter between the core
```

Comment:
I don't mind switching the dev version naming to use hyphens before "dev" moving forward.

Reply:
I don't think really think the difference between `.devN` and `-devN` is a big deal (it's just a very minor detail...). I'm happy to go either way.

However, what would you think about slightly altering the way that we determine the numbers for the next `dev` version.
- currently, we always increment the minor version. (So after we release `3.3.0`, we will create `3.4-dev1`)
    - I think `yt` does this. That works well for them since the project maintains separate branches for prior minor-release numbers and micro-releases correspond to various commits along those minor-release branches.
- what it instead, we always increment the micro version by default (after we release `3.3.0` we would create `3.3.1-dev`)
    - I think this might work better for us since we only maintain a single branch and a release corresponds to a single point along that branch.


**Pros and Cons:**
- Advantage: it simplifies process of specifying forward-compatible version requirements. An example helps illustrate this. Consider a hypothetical series of releases: `3.3.0`, `3.3.1`, `3.3.2`, `3.3.3`, `3.4.0`, ...
   - Now imagine, that some downstream-simulation code developer encountered some critical bug in the `3.3.0` and `3.3.1` releases. They then submit a upstream a patch to fix the bug, and they want to add some check to their build-system (or even just a note in their docs) to prevent their users from compiling with the known incompatible version. What do they do?
   - Under the current approach, there isn't a simple answer. Under this approach, the full ordered set of version numbers might include `3.3.0`, `3.4.dev1`, `3.3.1`, `3.4.dev2`, `3.3.2`, `3.4.dev3`, `3.3.3`, `3.4.dev4`, `3.4.0`. The downstream developer needs to require that the leading numbers start with `3.3.2` (or newer) AND that the version-string is not `3.4.dev1`
    - Under my proposed approach, this is simpler. Under this approach, the full ordered set of version numbers might include `3.3.0`, `3.3.1-dev`, `3.3.1`, `3.3.2-dev`, `3.3.2`, `3.3.3-dev`, `3.3.3`, `3.3.4-dev`, `3.4.0`. The downstream developer must simply require that the version number starts with `3.3.2` or newer
- Disadvatnage: we might have versions in the order `3.3.0`, `3.3.1-dev`, `3.4.0`. While this might be a little confusing, it makes sense if you think of the leading numbers of a `-dev` version as a lower-bound on the next release (plus it would be trivial to increment the minor-number of the `-dev` version once new functionality is introduced to the main-branch that will clearly require a minor-release)

The benefits are a little contrived here. So it may be better to stick with what we were doing for the sake of consistency. What do you think?

Reply:
Yes, I like it. I've tried in the past to make use of the dev number, but mostly arbitrarily based on when big PRs landed. Let's go with your proposal.

Reply:
I think your approach works better if we were doing something more like yt where:
- we maintained separate branches corresponding to each minor release (e.g. a `3.2.x` branch)
- we backported changes from the main development branch to minor-release branch and then performed releases off of the minor-release branch.

In principle that sounds like a more robust solution to me with the most flexibility. But it also sounds like a tremendous amount of work for a grackle-sized project without much payoff...